compiler/rustc_typeck/src/check/expr.rs

```diff
@@ -1030,6 +1031,36 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
         let t_expr = self.check_expr_with_expectation(e, ExpectCastableToType(t_cast));
         let t_cast = self.resolve_vars_if_possible(t_cast);
 
+        // ptr_to_int_cast lint: when casting a function pointer to a numeric type it only makes
+        // sense to cast to `usize`. Other cases are often bugs. See #81686 as an example.
+        if t_expr.is_fn() || t_expr.is_raw_ptr() {
+            let to_numeric = matches!(t_cast.kind(), ty::Int(_) | ty::Uint(_) | ty::Float(_));
```

Comment:
```suggestion
            let is_integral_cast = matches!(t_cast.kind(), ty::Int(_) | ty::Uint(_));
```
It's not possible to cast pointers to floats anyway.